### PR TITLE
Define all prims when exporting transform hierarchies (#USDU-217)

### LINKED ISSUE #278: USD export from Unity3D not respecting hierarchical transformations (#USDU-217)
[](url)I have a few scenes in Unity that I am trying to export using the USD package. That package has various options for export (from the Menu, over the recorder, via API calls), and I have tried them all, but the resulting file (inspected with usdview) – while it has all the graphics pieces inside - doesn’t respect the hierarchical transformations that I have set up.

For more details see this link SO:

[https://stackoverflow.com/questions/70246077/usd-export-from-unity3d-not-respecting-hierarchical-transformations](url)

Happy to discuss it here, but I am cross-posting to a few places and SO seemed most central.

### PR DESCRIPTION
**Pull request by judubu**
## Purpose of this PR

**Ticket/Jira #:** https://jira.unity3d.com/browse/USDU-217

This fixes a regression introduced with 3.0.0-exp.2

The current scene exporter writes transform hierarchies bottom up and the serialization logic only checks the prim validity before serializing.
But when writing a prim at the bottom of a hierarchy USD defines automatically all the prims of the hierarchy. This leads to typeless transform hierarchies.

This PR make sure the serialization code only write to prims whose type match the schema of the sample being serialized.

Fixes #278 

## Testing

**Functional Testing status:**

Create a hierarchy of 2 empty game objects and a cube, export.

**Before**
```
def "GameObject"
{
    matrix4d xformOp:transform = ( (1, 0, 0, 0), (0, 1, 0, 0), (0, 0, 1, 0), (161.81143188476562, 65.73773193359375, 88.41531372070312, 1) )
    uniform token[] xformOpOrder = ["xformOp:transform"]

    def "GameObject__1_"
    {
        matrix4d xformOp:transform = ( (1, 0, 0, 0), (0, 1, 0, 0), (0, 0, 1, 0), (1, 0, 0, 1) )
        uniform token[] xformOpOrder = ["xformOp:transform"]

        def Mesh "Cube"
        {
            .....
            matrix4d xformOp:transform = ( (1, 0, 0, 0), (0, 1, 0, 0), (0, 0, 1, 0), (6, 0, 0, 1) )
            uniform token[] xformOpOrder = ["xformOp:transform"]
        }
    }
}
```

**After**
```
def Xform "GameObject"
{
    matrix4d xformOp:transform = ( (1, 0, 0, 0), (0, 1, 0, 0), (0, 0, 1, 0), (161.81143188476562, 65.73773193359375, 88.41531372070312, 1) )
    uniform token[] xformOpOrder = ["xformOp:transform"]

    def Xform "GameObject__1_"
    {
        matrix4d xformOp:transform = ( (1, 0, 0, 0), (0, 1, 0, 0), (0, 0, 1, 0), (1, 0, 0, 1) )
        uniform token[] xformOpOrder = ["xformOp:transform"]

        def Mesh "Cube"
        {
            .....
            matrix4d xformOp:transform = ( (1, 0, 0, 0), (0, 1, 0, 0), (0, 0, 1, 0), (6, 0, 0, 1) )
            uniform token[] xformOpOrder = ["xformOp:transform"]
        }
    }
}
```

This PR adds a test checking that prim types match sample types after serialization.

## Overall Product Risks

**Complexity:**

Minimal

**Halo Effect:**

Low

### CODE REVIEW

**Review by cguer — APPROVED** [review]
QA approved 👍!

Tested on:
- 2021.2.7f1 (Windows only)

File: package/com.unity.formats.usd/Dependencies/USD.NET/serialization/Scene.cs [hunk 1]
```diff
@@ -793,19 +793,18 @@ public void Read<T>(SdfPath path, System.Reflection.PropertyInfo propInfo, ref T
...
+                    var primTypeName = Reflect.GetSchema(typeof(T));
+
+                    prim = m_stage.GetPrimAtPath(path);
+                    if (prim == null || !prim.IsValid())
                     {
-                        prim = m_stage.DefinePrim(path, new TfToken(schema));
+                        prim = m_stage.DefinePrim(path, new TfToken(primTypeName));
                     }
-                    else
+                    else if (!string.IsNullOrEmpty(primTypeName) && (string.IsNullOrEmpty(prim.GetTypeName().GetText()) || primTypeName != "Xform"))
```
Comment (vickycl) [inline package/com.unity.formats.usd/Dependencies/USD.NET/serialization/Scene.cs:805]: Is it valid to create a TfToken for a null or empty string? If not, I believe the first string check should be done earlier, at the point of getting primTypeName before defining the prim in the other branch of the if

File: package/com.unity.formats.usd/Tests/USD.NET/SceneTests.cs [hunk 2]
```diff
@@ -119,5 +119,66 @@ public void WritePathToSceneFile_WithInvalidPath_ThrowsException(string path)
             Assert.Throws<System.Exception>(() => scene.Write(path, new SampleBase()));
             UnityEngine.TestTools.LogAssert.Expect(LogType.Exception, string.Format("ApplicationException: USD ERROR: Path must be an absolute path: <{0}>", path));
         }
+
+        [Test]
+        public static void StartEndTimeTest()
```
Comment (vickycl) [inline package/com.unity.formats.usd/Tests/USD.NET/SceneTests.cs:124]: Minor comment: As these are being moved could they also be given names that match the test name format followed by the other tests in this file?

/ I'm not sure whether there's much value in testing the assignment of a value, but maybe I'm missing something that can go wrong here :)

**Review by vickycl — COMMENTED** [review]
Added a couple of questions

File: package/com.unity.formats.usd/Dependencies/USD.NET/serialization/Scene.cs (hunk 1 above)
Comment (vickycl) [reply package/com.unity.formats.usd/Dependencies/USD.NET/serialization/Scene.cs:805]: Discussed on Slack with Mircea- the primTypeName will not be null and empty is a valid type to define a prim with, so this check is ok

File: package/com.unity.formats.usd/Tests/USD.NET/SceneTests.cs (hunk 2 above)
Comment (mirceaispas) [reply package/com.unity.formats.usd/Tests/USD.NET/SceneTests.cs:124]: I think this test was moved from StageTests.cs and I assume this file suits better the test. 
The actual relevant tests for this bug are in `class SceneWriteTests : UsdTests` which define objects in a reverse order (child Prims get defined before parents which will create typeless Prims)

**Review by vickycl — APPROVED** [review]

**Pull request merged**